# JBiclustGE

Please visit [website](https://jbiclustge.github.io/) with more detailed information.

Is an application programming interface (API) developed in the Java language, that comprises 21 biclustering methods to be used in the context of gene expression data analysis. Currently, this application integrates the following biclustering algorithms:

* [Bayesian Biclustering Model](http://www.people.fas.harvard.edu/~junliu/BBC/) (BBC)
* [BicFinder](http://www.info.univ-angers.fr/pub/hao/BicFinder.html)
* [BiMine+](http://www.info.univ-angers.fr/pub/hao/BiMine+/BiMine+.html)
* [BIclustering by Correlated and Large number of Individual Clustered seeds](http://scode.kaist.ac.kr/BICLIC/) (BICLIC)
* Plaid, Spectral, xMOTIFS, Cheng & Church using the R library [biclust](https://cran.r-project.org/web/packages/biclust/) as backend. 
* [Bit-Pattern Biclustering Algorithm](http://eps.upo.es/bigs/BiBit_algorithm.html) (BiBit)
* FLexible Overlapped biClustering (FLOC) using R library [Bicare](https://www.bioconductor.org/packages/release/bioc/html/BicARE.html) as backend. 
* [Combinatorial Algorithm for Expression and Sequence-based Cluster Extraction](https://bitbucket.org/libsleipnir/sleipnir) (COALESCE)
* [Correlated Pattern Biclusters](http://bmi.osu.edu/hpc/software/cpb/index.html) (CPB)
* [Differentially Expressed BIclusters](https://github.com/razZ0r/debi) (DeBi)
* [Factor Analysis for Bicluster Acquisition](https://www.bioconductor.org/packages/release/bioc/html/fabia.html) (FABIA)
* [Iterative signature algorithm](https://www.bioconductor.org/packages/release/bioc/html/fabia.html) (ISA)
* [Order-preserving sub-matrices](http://www.tik.ee.ethz.ch/sop/bicat/) (OPSM)
* [BiMax](http://www.tik.ee.ethz.ch/sop/bimax/)
* [Penalized Plaid](http://www.dms.umontreal.ca/~murua/)
* [QUalitative BIClustering](https://github.com/maqin2001/qubic) (QuBic)
* [UniBic](https://sourceforge.net/projects/unibic/files/)
* [UBCLUST](http://alumni.cs.ucr.edu/~hli/ubc/)



This application also provides additional features such as:

* Data preprocessing (Normalization, Standardization, Binarization, Missing values imputation).
* Parallel execution of the algorithms (multithread attributes).
* Built-in frameworks for the Gene Set Enrichment Analysis (GSEA).
* Evaluation measures.
* Generation of synthetic datasets.
* Visualization of biclustering results (Heatmaps and Parallel coordinates).  


